lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -538,6 +546,80 @@ ArrayAttr TypeLoweringVisitor::filterAnnotations(MLIRContext *ctxt,
   return ArrayAttr::get(ctxt, retval);
 }
 
+LogicalResult TypeLoweringVisitor::partitionSymbols(
+    hw::InnerSymAttr sym, ArrayRef<FlatBundleFieldEntry> fields,
+    SmallVectorImpl<hw::InnerSymAttr> &newSyms, Location errorLoc) {
+
+  // No symbol, nothing to partition.
+  if (!sym)
+    return success();
+
+  auto *context = sym.getContext();
+
+  // Helper for getting max fieldID from const FIRRTLBaseType's.
+  auto getMaxFieldID = [](const FIRRTLBaseType type) {
+    return FIRRTLBaseType(type).getMaxFieldID();
+  };
+
+  newSyms.resize(fields.size());
+
+  // Sort inner symbols by fieldID if not already sorted.
+  auto props = llvm::to_vector(sym);
+  if (props.empty())
+    return success();
+  llvm::stable_sort(props, [&](const auto &lhs, const auto &rhs) {
+    return lhs.getFieldID() < rhs.getFieldID();
+  });
+
+  // Double-check fields are in increasing order.
+  assert(llvm::is_sorted(fields, [&](const auto &lhs, const auto &rhs) {
+    return lhs.fieldID < rhs.fieldID &&
+           (rhs.fieldID - lhs.fieldID) > getMaxFieldID(lhs.type);
+  }));
+
+  // Walk fields, gather symbols that target each and assign with
+  // fieldID's updated to be relative.
+  // If any symbols target fieldID's not covered by these ranges,
+  // they cannot be preserved and generate an error.
+  auto propIt = props.begin(), propEnd = props.end();
+  for (auto [field, newSym] : llvm::zip(fields, newSyms)) {
+    // Error if inner symbol target is before this.
+    // Exit and let tail code produce diagnostic.
+    if (propIt->getFieldID() < field.fieldID)
+      break;
+
+    // Consume inner_syms on this field.
+    SmallVector<hw::InnerSymPropertiesAttr> propsForField;
```

Comment:
The rootChildFieldID is suppose to make the pealing process easy as it let's you rebase a fieldID when you split the root type once.

Reply:
Alternately, getIndexForFieldID should give the root index and the adjusted fieldID for any fieldID.
I guess I'm suggesting first iterating on the symbols and adjusting them before sorting them into fields for later.  And also not reinventing the adjustment code which is already in FIRRTLType.

Reply:
Refactored to use `rootChildFieldID`!

Unsure what you mean re:adjusting them before sorting them into fields, sorry.  Adjusting them probably needs to be tied to some sort of figuring out what field they go to (otherwise what are you adjusting relative to, and how do you know later which it belongs to if not assigning them as part of process doing adjustment)?

FWIW this does do this walk /once/ per lowering layer, vs annotation filtering which for each element scans all annotations, so maybe it already does what you meant?

Reply:
Okay, reworked to bin explicitly and then build per-index syms!